pkg/relay/ethereum/tx.go

```diff
@@ -460,3 +429,265 @@ func (c *Chain) parseRpcError(err error) (string, string) {
+			return nil, err
+		}
+
+		txGasLimit, err := estimateGas(ctx, c, tx, true, logger)
+		if err != nil {
+			return nil, err
+		}
+		opts.GasLimit = txGasLimit
+	}
+
+	opts.NoSend = false
+	tx, err := c.SendTx(opts, *iter.Current(), iter.skipUpdateClientCommitment)
+	if err != nil {
+		logger.Error("failed to send msg", err)
+		return nil, err
+	}
+	iter.Next()
+	return tx, nil
+}
+
+func (iter *CallIter) sendMultiTx(ctx context.Context, c *Chain) (*gethtypes.Transaction, error) {
+	if (iter.End()) {
+		return nil, nil
+	}
+	// now iter.cursor < len(iter.msgs)
+
+	logger := c.GetChainLogger()
+
+	opts, err := c.TxOpts(ctx, true);
+	if err != nil {
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if iter.txs == nil { // create txs at first multicall call
+		opts.GasLimit = math.MaxUint64
+		opts.NoSend = true
+		txs := make([]gethtypes.Transaction, 0, len(iter.msgs))
+		msgTypeNames := make([]string, 0, len(iter.msgs))
+
+		for i := 0; i < len(iter.msgs); i++ {
+			msgTypeName := fmt.Sprintf("%T", iter.msgs[i])
+			msgTypeNames = append(msgTypeNames, msgTypeName)
+
+			tx, err := c.SendTx(opts, iter.msgs[i], iter.skipUpdateClientCommitment)
+			if err != nil {
+				logger := &log.RelayLogger{Logger: logger.With(
+					logAttrMsgIndexFrom, i,
+					logAttrMsgIndexTo, i,
+					logAttrMsgType, msgTypeName,
+				)}
+				logger.Error("failed to build tx for gas estimation", err)
+				return nil, err
+			}
+			if tx.To() == nil {
+				err2 := fmt.Errorf("no target address")
+				logger.Error("failed to construct Multicall3Call", err2)
+				return nil, err2
+			}
+			txs = append(txs, *tx)
+		}
+		iter.txs = txs
+		iter.msgTypeNames = msgTypeNames
+	}
+
+	var (
+		lastOkCalls []multicall3.Multicall3Call = nil
+		lastOkGasLimit uint64 = 0
+	)
+	count, err := findItems(
+		len(iter.msgs) - iter.Cursor(),
+		func(count int) (error) {
+			from := iter.Cursor()
+			to := from + count
+
+			logger := &log.RelayLogger{Logger: logger.With(
+				logAttrMsgIndexFrom, from,
+				logAttrMsgIndexTo, from + count - 1,
+				logAttrMsgType, strings.Join(iter.msgTypeNames[0 : from + count], ","),
+			)}
+
+			calls := make([]multicall3.Multicall3Call, 0, count)
+			for i := from; i < to; i++ {
+				calls = append(calls, multicall3.Multicall3Call{
+					Target: *iter.txs[i].To(),
+					CallData: iter.txs[i].Data(),
+				})
+			}
+
+			opts.GasLimit = math.MaxUint64
+			opts.NoSend = true
+			multiTx, err := c.multicall3.Aggregate(opts, calls)
+			if err != nil {
+				return err
+			}
+
+			txGasLimit, err := estimateGas(ctx, c, multiTx, 1 == count, logger)
+			if err != nil {
+				return err
+			}
+
+			lastOkGasLimit = txGasLimit
+			lastOkCalls = calls
+			return nil
+		})
+
+	logger = &log.RelayLogger{Logger: logger.With(
+		logAttrMsgIndexFrom, iter.Cursor(),
+		logAttrMsgIndexTo, iter.Cursor() + count - 1,
+		logAttrMsgType, strings.Join(iter.msgTypeNames[0 : iter.Cursor() + count], ","),
+	)}
+
+	if err != nil {
+		logger.Error("failed to multicall", err)
```

Comment:
@dai1975 This message a bit too simple. Can we rewrite it like "failed to prepare multicall tx"?

Reply:
modified.